src/Dsll.hs

```diff
@@ -0,0 +1,397 @@
+-- | "DSLL" contains the grammers and parser for the 
+--    DSLL language
+--    Author: Dor Ma'ayan, May 2018
+
+{-# OPTIONS_HADDOCK prune #-}
+--module DSLL where
+module Main (main) where -- for debugging purposes
+
+import System.Process
+import Control.Monad (void)
+import Data.Void
+import Debug.Trace
+import System.IO -- read/write to file
+import System.Environment
+import Control.Arrow ((>>>))
+import System.Random
+import Debug.Trace
+import Data.Functor.Classes
+import Data.List
+import Data.Maybe (fromMaybe)
+import Data.Typeable
+import Text.Megaparsec
+import Text.Megaparsec.Char
+import Text.Megaparsec.Expr
+--import Text.PrettyPrint
+--import Text.PrettyPrint.HughesPJClass hiding (colon, comma, parens, braces)
+import qualified Data.Map.Strict as M
+import qualified Text.Megaparsec.Char.Lexer as L
+--------------------------------------- DSLL AST ---------------------------------------
+type TypeVar = String
+
+type Var = String
+
+type Prop = String
+
+type Type = String
+
+
+data ConstructorInvoker = ConstructorInvoker { nameCons :: String, argCons:: [Arg]}
+    deriving (Eq, Typeable)
+instance Show ConstructorInvoker where
+    show (ConstructorInvoker nameCons argCons) = nString ++ "(" ++ aString ++ ")"
+        where nString = show nameCons
+              aString = show argCons
+
+
+data T = TTypeVar TypeVar --TODO: figure out a better name for that...
+    | TConstr ConstructorInvoker
+    deriving (Show, Eq, Typeable)
+
+data Arg = AVar Var
+    | AT T
+    deriving (Show, Eq, Typeable)
+
+data K = Ktype Type --TODO: figure out a better name for that...
+    | KT T
+    deriving (Show, Eq, Typeable)
+
+data Cd = Cd{nameCd :: String, inputCd::[K], outputCd::Type}
+    deriving (Eq, Typeable)
+
+instance Show Cd where
+    show (Cd nameCd inputCd outputCd) = "(TCon, " ++ nString ++ " ,ValOfType " ++ iString ++ "Output " ++ oString ++")"
+        where nString = show nameCd
+              iString = show inputCd
+              oString = show outputCd
+        
+
+data Od = Operator String [Var] [T] [TypeVar] [Type] [T] T 
+    deriving (Show, Eq, Typeable)
+    
+-- data Od = Od{ nameOd :: String, forVarsoD :: [Var], typesForVarsOd :: [T], forTypesOd :: [TypeVar], typeForTypesOd :: [Type]
+--               ,fromOd:: [T], toOd:: T}
+--     deriving (Show,Eq, Typeable)
+
+-- -- instance Show Od where
+-- --     show (Od nameOd forVarsoD typesForVarsOd forTypesOd typeForTypesOd fromOd toOd) = 
+-- --     	"(Op, " ++ nString ++ " ,ValOfType " ++ iString ++ "Output " ++ oString ++")"
+-- --         where aString = show nameOd
+-- --               bString = show inputCd
+-- --               cString = show outputCd
+-- --               dString = 
+-- --               eString = 
+
+
+data Pd = Predicate String [Var] [T] [TypeVar] [Type] [T] [Prop] Prop 
```

Comment:
Probably better to have Od and Pd as record types too (looks like you're already converting them)